### PR TITLE
Zopkio junit reporter

### PR DESCRIPTION
**Pull request by chuongddh-zz**
Hello there,

We adapted Zopkio for our Jenkins integration test. Therefore, we created another reporter: junit_reporter which will generate test report in Junit format using junit_xml package. Jenkins will then pick up report and mark build passed/failed accordingly.

Thanks,
Chuong

### CODE REVIEW

**Comment by arpras** [comment]
Hi Chuong,

Thanks for submitting the pull request. can you take a look at why the continous integration with travis-ci is failing with these changes ? We need it to be clean before merging these changes. Let me know if you need any help in looking at the failure.

Thanks

**Comment by jdehrlich** [comment]
Also can you rebase and reduce the number of commits to highlight important changes?
Thanks,
Joshua

File: zopkio/test_runner.py [hunk 1]
```diff
@@ -32,7 +32,8 @@
 
 import zopkio.constants as constants
 import zopkio.error_messages as error_messages
-from zopkio.reporter import Reporter
+# from zopkio.reporter import Reporter
```
Comment (jdehrlich) [inline zopkio/test_runner.py]: please delete rather than commenting out

File: zopkio/test_runner.py [hunk 2]
```diff
@@ -32,7 +32,8 @@
...
+from reporters import junit_reporter, html_reporter
```
Comment (jdehrlich) [inline zopkio/test_runner.py]: please use the qualified import (zopkio.reporters)

File: zopkio/test_runner.py [hunk 3]
```diff
@@ -77,17 +78,22 @@ def __init__(self, *args, **kwargs):
     :param kwargs:
     :return:
     """
+    # if ('reporter_type' in kwargs):
```
Comment (jdehrlich) [inline zopkio/test_runner.py]: please delete these lines

File: zopkio/test_runner.py [hunk 4]
```diff
@@ -157,10 +163,11 @@ def run(self):
               if not setup_fail:
                 failure_handler.notify_failure()
               logger.error("{0} failed teardown_suite(). {1}".format(config.name, traceback.format_exc()))
-        finally:
```
Comment (jdehrlich) [inline zopkio/test_runner.py]: I think we still want this in a finally block?

File: zopkio/reporters/junit_reporter.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
+# Copyright 2014 LinkedIn Corp.
```
Comment (jdehrlich) [inline zopkio/reporters/junit_reporter.py]: change year? Awesome

Comment (chuongddh) [inline zopkio/reporters/junit_reporter.py]: It was copied from html_reporter file. I dont think i change the year there.

Chuong

**Comment by chuongddh** [comment]
Hi Joshua and Arpras,

I will review these change later when I have a chance.

Thanks,
Chuong

**Comment by chuongddh** [comment]
Hi Arpras,

I have taken a look, it seems the build failed due o 2 reason:

1.  I made some small change to test junit_reporter which may be failing
the test, because I just wanted to validated the test result report.
2. This Junit reporter is using Python package junit_xml, which need to be
install in the environment which is using Zopkio.

As a side note, by default it still uses html reporter. Only when we use
the variable --reporter=junit_reporter, it will generate xml for Junit
format test report. In addition, I know that I have not clean up the code;
I will try look into it next in next couple week and submit another pull
request.

Kind Regards,
Chuong

**Comment by arpras** [comment]
Thanks Chuong. This is really a cool feature and we really appreciate the contribution in making Zopkio better.

Thanks.

**Comment by arpras** [comment]
Hi Chuong,

Did you get a chance to look at this ?

Thanks.

**Comment by chuongddh** [comment]
Hi Arpras,

Thanks for your email. I have not had chance to clean it up. I'm currently
out of office for family matter.

Kind Regards,
Chuong

**Comment by arpras** [comment]
Hi Chuong,

Thanks. Let me work on cleaning this up then. This seems to be a popular feature request and I will try to work on this this month and send out a pull request for review. Will include you on the pull request. Please review if you get time.

Thanks

**Comment by chuongddh** [comment]
Hi Arpras,

Thank you so much. I just got back to office after maternity leave. And I'm
sorry, I was putting something together in short time for that pull
request. The change was simple, with a new reporter file for junit using
junit_xml. I also made change for some quick test which may be why it
failed your build.

I just picked up Zopkio again to use ZTestSuite and ZTest, trying to extend
the python tests with yml file structure. I will try to review the pull
request, you can let me know what you would like to change.

By the way, do you have any document on using ZTest and ZTestSuite with
Zopkio?

Thanks and Regards,
Chuong

**Comment by arpras** [comment]
Congratulations :)

Thanks for the interest in ZTest. We have some examples and much of the documentation is built into the code. examples : Zopkio/examples/zookeeper/zookeeper_ztestsuite_example.py

Thanks

**Comment by chuongddh** [comment]
Thank you!

-Chuong

**Comment by chuongddh** [comment]
Hi Arpras,

I have pull changes from Zopkio's master and made changes to my branch according to above comments. Can you take a look at the pull request; and let me know if there is any problem?

I have not added any unit test for this junit reporter, however, we have been using it extensively here in Pubnub. The syntax for it is as below. It will generate junit xml report file in the reporting folder.

$zopkio zopkio_test.py --reporter=junit_reporter  

Kind Regards,
Chuong

**Comment by arpras** [comment]
Hi Chuong,

Thanks for fixing some of the issues. I still see that the changes fail the travis build. Other issues I see are that passing from command line seems to involve more intrusive code changes. There are also 30 commits for this one pull request and it is probably not a good idea.

As we discussed before I made some changes to your submission and submitted a pull request 101. Can you review if it looks good ? It passes travis build and also carries unit tests. Also the way to switch to junit runner is using configs so the code structure and changes to test runner/zopkio main are very limited. Please let me know what you think. 

Thanks

**Comment by arpras** [comment]
Also for some reason github does not allow me to add my changes to the same pull request (probably because the underlying branch origin is different (from pubnub)). So I had to create a new pull request #101 . 

Thanks

**Comment by arpras** [comment]
Thanks for reviewing the pull request chuong. 

If you are comfortable with it can I add in comments that this was "done in collaboration with Chuong of PubNub". Let me know if you are ok with this, then I will make the changes.

Also  would you be able to join zopkio@googlegroups.com. This would be a great way to collaborate and plan on say having a joint meeting session for zopkio where we can together present its features and describe how we are using it.

Thanks,
Arun

**Comment by chuongddh** [comment]
Hi Arpras,

Thanks for the details. I looked at request #101, it is fine with me. And
yes, I'm good with using config for reporter configuration.

By the way, I could not find Zopkio google group. Is it an open group that
I can join?

Thanks,
Chuong

**Pull request closed without merging**